### PR TITLE
Async recipes

### PR DESCRIPTION
**Pull request by mweststrate**
Localized branch of #305, implements #302 

Remaining work: 

* [x] docs
* [x] `createDraft`  / `finishDraft` api
* [x] typings `create`/`finish` `Draft`
* [x] typings async produce

### CODE REVIEW

**Comment by aleclarson** [comment]
Heads up, I force-pushed to remove the merge commit and reword the createDraft/finishDraft commit so it shows up on the Releases page.

**Comment by mweststrate** [comment]
Oww sorry, yes I should read up on the new commit / release process. Any
recommended link?

Op za 2 feb. 2019 14:38 schreef Alec Larson <notifications@github.com:

**Comment by aleclarson** [comment]
Here you go: https://www.conventionalcommits.org/en/v1.0.0-beta.2/ 👍

**Comment by aleclarson** [comment]
I wonder if `createDraft` should "tag" its drafts so we can throw if a `produce`-made draft is ever passed to `finishDraft`.

**Comment by mweststrate** [comment]
Good point, Wii add

Op za 2 feb. 2019 15:00 schreef Alec Larson <notifications@github.com:

**Comment by aleclarson** [comment]
Also, the types for `produce` require no changes. 🎉

**Comment by mweststrate** [comment]
Not for the async version? I would expect to need something like
produce<T>(Draft<T> => Promise<void | T>): Promise<T> ?

Op za 2 feb. 2019 om 15:11 schreef Alec Larson <notifications@github.com>:

**Comment by aleclarson** [comment]
Ah good point, didn't think of the `Promise<void>` case. Would that make this a breaking change?

**Comment by aleclarson** [comment]
I'm updating the tests to use snapshots for error messages and patches. 😎

**Comment by aleclarson** [comment]
We may want to consider changing the return type of `finishDraft` to `Immutable<T>`, because TypeScript cannot infer `T` from `Draft<T>`. Thus, the current return type of `finishDraft` is mutable.

**Comment by aleclarson** [comment]
Okay, the return type of `produce` has been updated and simplified! 🚀

**Comment by mweststrate** [comment]
@aleclarson awesome! I'll proceed with the docs in an iffy, and then we'll should be ready to merge I guess!

**Comment by aleclarson** [comment]
Since we'll be publishing a major, we must decide if https://github.com/mweststrate/immer/issues/246 should be addressed, too.

**Comment by mweststrate** [comment]
Eh.. why no minor?

Op za 2 feb. 2019 19:02 schreef Alec Larson <notifications@github.com:

**Comment by aleclarson** [comment]
As mentioned [here](https://github.com/mweststrate/immer/pull/307#issuecomment-459968964), this PR causes `produce` to return `Promise<string | T>` where it previously returned `Promise<string | undefined>`.

**Comment by mweststrate** [comment]
But that only changes the static types, not the runtime behavior, correct?
(Imho the best practice is not to count these as breaking changes, cause
otherwise every time the typings improve, you have a breaking change, while
usually it just reveals an already wrong assumption on the consumer)

Op za 2 feb. 2019 19:05 schreef Alec Larson <notifications@github.com:

**Comment by aleclarson** [comment]
We never said in the docs that returning a promise from a producer is not supported. We just said not to modify the draft in an "async process", which isn't the same thing.

> But that only changes the static types, not the runtime behavior, correct?

Nope. Runtime behavior is affected, because we're changing the result of the returned promise. Promises that previously resolved with an undefined value will now resolve with the base state or a modified copy.

**Comment by mweststrate** [comment]
Ok check! I think #246 can be done separately, as that one requires a real
migration

Op za 2 feb. 2019 19:16 schreef Alec Larson <notifications@github.com:

File: readme.md [hunk 1]
```diff
@@ -335,6 +335,46 @@ For a more in-depth study, see [Distributing patches and rebasing actions using
...
+```javascript
+import { createDraft, finishDraft } from "immer"
+
+const user = {
+    name: "michel",
+    todos: []
+}
+
+const draft = createDraft(user)
+const draft.todos = await window.fetch("http://host/" + draft.name).json()
```
[the remark this answers is not in the document]
Comment (mweststrate) [reply readme.md]: fixed

**Comment by aleclarson** [comment]
You stole all the credit by squashing it. 😓

**Comment by mweststrate** [comment]
Darn, thought that was required for semantic-release. If you want I
force-push a normal merge.

Op za 2 feb. 2019 20:55 schreef Alec Larson <notifications@github.com:

**Comment by mweststrate** [comment]
Done, an cancelled the resulting build. Hopefully this doesn't mess up the next semantic release computation :-P.

**Comment by aleclarson** [comment]
If it does, I think we just need to manually push the next version to fix it. Thanks! 👍

**Comment by mweststrate** [comment]
Good point, I'll release a no-op 2.0.1 to make sure it won't be forgotten

**Comment by mweststrate** [comment]
I force pushed a new tag `v2.0.0`, I'm pretty sure that should already fix it, release overview now links to the correct commit as well

**Pull request merged**